Nu/Nu/Ecs/EcsPrelude.fs

```diff
@@ -45,32 +46,66 @@ type Store =
         end
 
 /// Stores components for an Ecs.
-type Store<'c when 'c : struct and 'c :> 'c Component> (name) =
+type Store<'c when 'c: struct and 'c :> 'c Component>(name) =
     let mutable arr = Array.zeroCreate<'c> Constants.Ecs.ArrayReserve
     member this.Length = arr.Length
     member this.Name = name
     member this.Item i = &arr.[i]
     member this.SetItem index comp = arr.[index] <- comp
     member this.ZeroItem index = arr.[index] <- Unchecked.defaultof<'c>
-    member this.Grow () =
+
+    member this.Grow() =
         let length = int (single (max arr.Length 2) * 1.5f)
         let arr' = Array.zeroCreate<'c> length
-        Array.Copy (arr, arr', arr.Length)
+        Array.Copy(arr, arr', arr.Length)
         arr <- arr'
-    member this.Read index count (stream : FileStream) =
+
+    member this.Write(stream: FileStream) =
+        let ba: byte array = Unsafe.As &arr
+        stream.Write ba
+        stream.Flush()
+        stream.Close()
+
+    member this.OpenWrite(stream: FileStream) =
+        let ba: byte array = Unsafe.As &arr
+        stream.Write ba
+
+    member this.Read index count (stream: FileStream) =
         let compSize = sizeof<'c>
         let comp = Unchecked.defaultof<'c> :> obj
         let buffer = Array.zeroCreate<byte> compSize
-        let gch = GCHandle.Alloc (comp, GCHandleType.Pinned)
-        try 
+        let gch = GCHandle.Alloc(comp, GCHandleType.Pinned)
+
+        try
             let mutable index = index
+
             for _ in 0 .. dec count do
-                stream.Read (buffer, 0, compSize) |> ignore<int>
-                Marshal.Copy (buffer, 0, gch.AddrOfPinnedObject (), compSize)
-                if index = arr.Length then this.Grow ()
+                stream.Read(buffer, 0, compSize) |> ignore<int>
+                Marshal.Copy(buffer, 0, gch.AddrOfPinnedObject(), compSize)
+
+                if index = arr.Length then
+                    this.Grow()
+
                 arr.[index] <- comp :?> 'c
                 index <- inc index
-        finally gch.Free ()
+        finally
+            gch.Free()
+
+    member this.ReadFrom index count (stream: FileStream) =
+            let compSize = sizeof<'c>
+            let count = count / compSize
```

Comment:
I don't understand this line. The count parameter is the number of components to load. Why would that be divided by the component's size? Also, the index parameter denotes the component index at which to start reading in case you thought that was the byte offset or something.

Reply:
Let's say, you reading ints and you have array of 256 bytes.
You can read no more than 256/4=64 ints from that array.

Using stream.Read you advance stream cursor sizeof<'T> every time you call it.
So actual count of iteration must be count divided by sizeof<'T>